File: static/js/api/membership.js
/**
 * Membership API client for sub-region membership requests and invitations.
 */

import { get, post, del } from './client.js?v=c49aed9';

/**
 * Request to join a sub-region
 * @param {string} regionId - The sub-region ID to join
 * @returns {Promise<Object>} The created membership request
 */
export async function requestMembership(regionId) {
    return post(`/communities/${regionId}/membership-requests`, {});
}

/**
 * List the current user's membership requests
 * @param {Object} [filter] - Optional filter
 * @param {string} [filter.status] - Filter by status
 * @param {string} [filter.type] - Filter by type ('request' or 'invitation')
 * @returns {Promise<Object>} List of membership requests
 */
export async function listMyRequests(filter = {}) {
    return get('/membership-requests', filter);
}

/**
 * Cancel a pending membership request
 * @param {string} requestId - The request ID to cancel
 * @returns {Promise<Object>} Cancellation result
 */
export async function cancelRequest(requestId) {
    return del(`/membership-requests/${requestId}`);
}

/**
 * Get details of a specific membership request
 * @param {string} requestId - The request ID
 * @returns {Promise<Object>} Request details with votes
 */
export async function getRequest(requestId) {
    return get(`/membership-requests/${requestId}`);
}

/**
 * List pending membership requests for regions where user is admin
 * @returns {Promise<Object>} List of pending requests
 */
export async function listPendingForAdmin() {
    return get('/membership-requests/admin');
}

/**
 * List pending membership requests for a specific region
 * @param {string} regionId - The region ID
 * @returns {Promise<Object>} List of pending requests
 */
export async function listPendingForRegion(regionId) {
    return get(`/communities/${regionId}/membership-requests`);
}

/**
 * Vote on a membership request
 * @param {string} requestId - The request ID
 * @param {boolean} approve - True to approve, false to reject
 * @returns {Promise<Object>} Vote result
 */
export async function voteOnRequest(requestId, approve) {
    return post(`/membership-requests/${requestId}/vote`, { vote: approve });
}

/**
 * Invite a user to join a sub-region
 * @param {string} regionId - The sub-region ID
 * @param {string} userId - The user ID to invite
 * @returns {Promise<Object>} The created invitation
 */
export async function createInvitation(regionId, userId) {
    return post(`/communities/${regionId}/invitations`, { user_id: userId });
}

/**
 * List pending invitations for the current user
 * @returns {Promise<Object>} List of invitations
 */
export async function listMyInvitations() {
    return get('/invitations');
}

/**
 * Respond to an invitation
 * @param {string} invitationId - The invitation ID
 * @param {boolean} accept - True to accept, false to decline
 * @returns {Promise<Object>} Response result
 */
export async function respondToInvitation(invitationId, accept) {
    return post(`/invitations/${invitationId}/respond`, { accept });
}

export default {
    requestMembership,
    listMyRequests,
    cancelRequest,
    getRequest,
    listPendingForAdmin,
    listPendingForRegion,
    voteOnRequest,
    createInvitation,
    listMyInvitations,
    respondToInvitation,
};

File: internal/models/membership.go
package models

import (
	"time"
)

// MembershipRequestType represents the type of membership request
type MembershipRequestType string

const (
	MembershipRequestTypeRequest    MembershipRequestType = "request"    // User requests to join
	MembershipRequestTypeInvitation MembershipRequestType = "invitation" // Admin invites user
)

// MembershipRequestStatus represents the status of a membership request
type MembershipRequestStatus string

const (
	MembershipRequestStatusPending   MembershipRequestStatus = "pending"
	MembershipRequestStatusApproved  MembershipRequestStatus = "approved"
	MembershipRequestStatusRejected  MembershipRequestStatus = "rejected"
	MembershipRequestStatusExpired   MembershipRequestStatus = "expired"
	MembershipRequestStatusCancelled MembershipRequestStatus = "cancelled"
)

// SubRegionMembershipRequest represents a request to join a sub-region
type SubRegionMembershipRequest struct {
	ID             string                  `json:"id" db:"id"`
	UserID         string                  `json:"user_id" db:"user_id"`
	RegionID       string                  `json:"region_id" db:"region_id"`
	ParentRegionID string                  `json:"parent_region_id" db:"parent_region_id"`
	RequestType    MembershipRequestType   `json:"request_type" db:"request_type"`
	Status         MembershipRequestStatus `json:"status" db:"status"`
	InitiatedBy    string                  `json:"initiated_by" db:"initiated_by"`
	CreatedAt      time.Time               `json:"created_at" db:"created_at"`
	ExpiresAt      time.Time               `json:"expires_at" db:"expires_at"`
	ResolvedAt     *time.Time              `json:"resolved_at,omitempty" db:"resolved_at"`
}

// SubRegionMembershipVote represents a vote on a membership request
type SubRegionMembershipVote struct {
	ID        string    `json:"id" db:"id"`
	RequestID string    `json:"request_id" db:"request_id"`
	VoterID   string    `json:"voter_id" db:"voter_id"`
	Vote      bool      `json:"vote" db:"vote"` // true=approve, false=reject
	VotedAt   time.Time `json:"voted_at" db:"voted_at"`
}

// CreateMembershipRequestRequest represents the request body for creating a membership request
type CreateMembershipRequestRequest struct {
	RegionID string `json:"region_id" validate:"required,uuid"`
}

// VoteMembershipRequestRequest represents the request body for voting on a membership request
type VoteMembershipRequestRequest struct {
	Vote bool `json:"vote"` // true=approve, false=reject
}

// CreateInvitationRequest represents the request body for inviting a user to a sub-region
type CreateInvitationRequest struct {
	UserID string `json:"user_id" validate:"required,uuid"`
}

// RespondToInvitationRequest represents the request body for responding to an invitation
type RespondToInvitationRequest struct {
	Accept bool `json:"accept"` // true=accept, false=decline
}

// MembershipRequestResponse represents the response after creating/voting on a request
type MembershipRequestResponse struct {
	RequestID         string    `json:"request_id"`
	UserID            string    `json:"user_id"`
	RegionID          string    `json:"region_id"`
	RequestType       string    `json:"request_type"`
	Status            string    `json:"status"`
	VotesNeeded       int       `json:"votes_needed"`
	CurrentVotes      int       `json:"current_votes"`
	ExpiresAt         time.Time `json:"expires_at"`
	YourVote          *bool     `json:"your_vote,omitempty"`
	MembershipGranted bool      `json:"membership_granted,omitempty"`
	Message           string    `json:"message,omitempty"`
}

// MembershipRequestListResponse represents the response for listing membership requests
type MembershipRequestListResponse struct {
	Requests []MembershipRequestSummary `json:"requests"`
}

// MembershipRequestSummary represents a summary of a membership request for listings
type MembershipRequestSummary struct {
	ID          string    `json:"id"`
	UserID      string    `json:"user_id"`
	Username    string    `json:"username"`
	RegionID    string    `json:"region_id"`
	RegionName  string    `json:"region_name"`
	RequestType string    `json:"request_type"`
	Status      string    `json:"status"`
	Votes       int       `json:"votes"`
	VotesNeeded int       `json:"votes_needed"`
	CreatedAt   time.Time `json:"created_at"`
	ExpiresAt   time.Time `json:"expires_at"`
}

// MembershipRequestDetailResponse represents full request details
type MembershipRequestDetailResponse struct {
	ID               string                 `json:"id"`
	UserID           string                 `json:"user_id"`
	Username         string                 `json:"username"`
	RegionID         string                 `json:"region_id"`
	RegionName       string                 `json:"region_name"`
	ParentRegionID   string                 `json:"parent_region_id"`
	ParentRegionName string                 `json:"parent_region_name"`
	RequestType      string                 `json:"request_type"`
	InitiatedBy      ProposalUser           `json:"initiated_by"`
	Status           string                 `json:"status"`
	VotesNeeded      int                    `json:"votes_needed"`
	CurrentVotes     int                    `json:"current_votes"`
	Votes            []MembershipVoteDetail `json:"votes"`
	CreatedAt        time.Time              `json:"created_at"`
	ExpiresAt        time.Time              `json:"expires_at"`
	ResolvedAt       *time.Time             `json:"resolved_at,omitempty"`
	CanVote          bool                   `json:"can_vote"`
	HasVoted         bool                   `json:"has_voted"`
}

// MembershipVoteDetail represents a vote with user details
type MembershipVoteDetail struct {
	VoterID  string    `json:"voter_id"`
	Username string    `json:"username"`
	Vote     bool      `json:"vote"`
	VotedAt  time.Time `json:"voted_at"`
}

// MembershipRequestListFilter represents filter options for listing membership requests
type MembershipRequestListFilter struct {
	Status      string // "pending", "approved", "rejected", "expired", "cancelled", or "" for all
	RequestType string // "request", "invitation", or "" for all
	RegionID    string
	UserID      string
}
